Хамидуллина Аниса

catsAndSuicide opened the pull request:
@xackill

catsAndSuicide committed:
created TagsCloudContainer and TagsCloudContainerTests projects, added infrastructure

catsAndSuicide committed:
added classes from TagsCloud1 task (layout algorithm)

catsAndSuicide committed:
added first interface implementations

catsAndSuicide committed:
changes in infrasrtucture

catsAndSuicide committed:
1) changes in implementation:

- added class Tag to contain info about tags in cloud (fontSize, location)
- SizesGenerator is replaced by FontSizeGetter
- fixed bug in ImageGenerator
- now DictionaryGenerator calculate not count, but frequency of cloud
 - added container using
- added class ImageSettings to contain info about output
2) added example of cloud

xackill commented:
**Общие замечания:**
1. Не забывай холиварить, если что-то непонятно или с чем-то не согласна :)
2. Разбиение по коммитам можно улучшить.
  <br/>**Что стоит выделять в отдельный коммит?**
  Ровно одно цельное (законченное) изменение. Один шаг к решению задачи: не слишком мелкий, и не слишком большой; способный уместиться в голове :) И, конечно, решение должно компилироваться. Оно не обязано работать корректно, но лучше, чтобы оно компилировалось.
  <br/>**Зачем?**
  \- Всегда можно быстро прочитать решение целиком: пробежаться по заголовкам коммитов, высокоуровнево провалидировать шаги. Удобно, когда это не занимает много времени;
  \- Всегда можно зачекаутить любой коммит и посмотреть его в ide. Поскольку они компилируемы, то по решению можно будет свободно ориентироваться;
  \- Когда в коммите нет "лишних" изменений (не связанных с текущим шагом) &mdash; намного проще искать нюансы и баги (как автору, так и ревьюеру). Проще через пару месяцев возвращаться к этому коммиту и понимать, почему код устроен так, а не иначе.
  <br/>**Что сейчас было не так? (немного примеров)**
  \- Проекты были добавлены в sln лишь в последнем коммите. Для ревью первых коммитов, приходилось добавлять csproj‑ы руками, чтобы посмотреть изменения в ide;
  \- Первый коммит подготавливает плацдарм, но законченного смысла не несет: непонятно как будут связаны компоненты. Правильная ли это связь? Забыто что-то? Этот коммит сложно валидировать;
  \- Последний коммит содержит практически все решение: значительное изменение инфраструктуры и di-контейнер и фикс бага и ... (ну, по коммит-мессаджу понятно =);
  \- Между ревью код в одном месте стоит менять ровно один раз: чтобы ревьюер не валидировал сначала одну инфраструктуру, а через пару коммитов &mdash; другую. Такой подход облегчит ревью (меньше смотреть), а также подготовит код к поддержке. Первая инфраструктура не прижилась, значит она не нужна истории (иначе последующие поколения будут тратить время на ее разбор, а чуть позже узнавать, что она не нужна).
  <br/>**Нужно ли сразу писать код на идеально нарезанные коммиты?**
  Нет. Скорее всего так не получится сделать. Просто после того, как закончишь решение, коммиты стоит прибрать при помощи `rebase -i` :)
  <br/>**Все ли так делают?**
  Совершенно точно нет; даже в рамках Контура. Кому-то удобно ревьюить диффы между первым и последним коммитом, кому-то при мерже в мастер склеивать ветку в один коммит. Однако я чаще сталкивался с тем, что выгоднее один раз потратить больше времени на этапе разработки (прибрав коммиты и оставив их для истории), чем потом тратить его каждый раз, разбираясь в этом коде (все-таки мы чаще читаем код, чем пишем).
  <br/>**Что я предлагаю сделать с этим замечанием?**
  Текущие коммиты можно не трогать; а вот последующие предлагаю делать по описанным выше правилам :)

xackill commented on TagsCloudContainer/Infrastructure/IFileReader.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    internal interface IFileReader
+    {
+        public IEnumerable<string> ReadLines(string filename);
```
Правда ли это любой поставщик слов зависит от fileName?
А если например слова читать из сети?
Кажется, над этой точкой расширения стоит подумать.

catsAndSuicide replied:
То есть лучше IFileReader превратить в более абстрактный IDataReader с методом ReadLines(), который ничего не принимает, а имя файла принимать в конструкторе конкретной реализации?

xackill replied:
Да, кажется так будет чище :)

xackill commented on TagsCloudContainer/Infrastructure/ICloudGenerator.cs:
```diff
@@ -1,10 +1,10 @@
     internal interface ICloudGenerator
     {
-        public Dictionary<string, Rectangle> GenerateCloud(Dictionary<string, int> dictionary);
+        public IEnumerable<Tag> GenerateCloud(Dictionary<string, double> frequencyDictionary, string fontName);
```
Правда ли fontName &mdash; ответственность метода GenerateCloud?
Может лучше сложить это в зависимость, связанную со шрифтами (которую уже сейчас принимает ICloudGenerator)?

catsAndSuicide replied:
Под зависимостью, связанную со шрифтами, подразумевается FontSizeGetter?

xackill replied:
Да, он. Возможно он будет как-то переосмыслен, так что я и говорил про что-то общее ("зависимость со шрифтами").

xackill commented on TagsCloudContainer/App/CloudGenerator/CloudGenerator.cs:
```diff
@@ -1,31 +1,38 @@
                 var word = pair.Key;
-                var size = pair.Value;
-                cloud[word] = layouter.PutNextRectangle(size);
+                var frequency = pair.Value;
```
Тут (и в похожих местах) можно изящно поиспользовать `ValueTuple` :)

xackill commented on TagsCloudContainer/App/CloudVisualizer/ImageGenerator.cs:
```diff
@@ -1,22 +1,20 @@
+        public Bitmap GenerateImage(IEnumerable<Tag> cloud, ImageSettings settings)
         {
-            using var image = new Bitmap(imageSize.Width, imageSize.Height);
+            var image = new Bitmap(settings.ImageSize.Width, settings.ImageSize.Height);
```
Баг поправлен :)
Но image теперь нигде не Dispose-ится :(

xackill commented on TagsCloudContainer/App/Program.cs:
```diff
@@ -1,9 +1,41 @@
+                .AddScoped<ICloudGenerator, CloudGenerator.CloudGenerator>()
+                .AddScoped<IImageGenerator, ImageGenerator>()
+                .AddSingleton<ICloudLayouter>(new CircularCloudLayouter(new Point(250, 250)))
+                .AddScoped<ICloudVisualizer, CloudVisualizer.CloudVisualizer>();
```
Почему какие-то зависимости `Scoped`, а какие-то `Singleton`? Отчего это зависит?

xackill commented on TagsCloudContainer/TagsCloudContainer.csproj:
```diff
@@ -1,8 +1,9 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>netcoreapp3.1</TargetFramework>
+    <UseWindowsForms>true</UseWindowsForms>
```
Неожиданно :)
Это из какого-то другого коммита?

xackill commented on TagsCloudContainer/App/FrequencyDictionaryGenerator/FrequencyDictionaryGenerator.cs:
```diff
@@ -2,35 +2,36 @@
         }
 
-        public Dictionary<string, int> GenerateDictionary(IEnumerable<string> lines)
+        public Dictionary<string, double> GenerateFrequencyDictionary(IEnumerable<string> lines)
```
Уже сейчас решение очень неплохо читается:  у тебя получилось уловить и реализовать ту логику разбиения, которая требуется di-контейнеру, отчего все части стали небольшими и достаточно понятными.

Но мне кажется, если чуть больше использовать LINQ, то получится еще читаемее.
Давай потренируемся в LINQ?
Например, попробуй реализовать этот метод по максимуму используя chaining и отказавшись от громоздких конструкций (где кстати происходит неоптимальная работа со словарем).

xackill commented on TagsCloudContainer/App/FileReader/TxtFileReader.cs:
```diff
@@ -0,0 +1,15 @@
+        public IEnumerable<string> ReadLines(string filename)
+        {
+            using var reader = new StreamReader(filename);
+            while (!reader.EndOfStream) yield return reader.ReadLine();
```
Ой, а есть способ это сделать проще?

xackill commented on TagsCloudContainer/App/Utils/RectangleExtensions.cs:
```diff
@@ -0,0 +1,50 @@
+            var rectCenter = new Point(
+                rectangle.X + rectangle.Width / 2,
+                rectangle.Y + rectangle.Height / 2);
+            return Math.Sqrt(Math.Pow(point.X - rectangle.X, 2) + Math.Pow(point.Y - rectCenter.Y, 2));
```
Тут правда требуется `rectangle.X`, а не `rectCenter.X`?

catsAndSuicide replied:
Упс, и вправду

xackill commented on TagsCloudContainer/App/Utils/DirectionUtils.cs:
```diff
@@ -0,0 +1,26 @@
+        public static DirectionToMove[] GetAllDirections()
+        {
+            return Directions;
+        }
```
А почему нельзя сделать массив сверху `public`?

xackill commented on TagsCloudContainer/App/CloudVisualizer/CloudVisualizer.cs:
```diff
@@ -19,12 +20,12 @@ internal class CloudVisualizer : ICloudVisualizer
-            var dictionary = dictionaryGenerator.GenerateDictionary(lines);
-            var cloud = cloudGenerator.GenerateCloud(dictionary);
-            return imageGenerator.GenerateImage(cloud, imageSize);
+            var dictionary = dictionaryGenerator.GenerateFrequencyDictionary(lines);
```
Еще про зависимости: не очень очевидно, но GenerateFrequencyDictionary парсит, нормализует и фильтрует слова, прежде чем посчитать их частоту. Надо бы это сделать очевиднее.

И кажется `dictionaryGenerator` не очень очевидное название.

xackill commented on TagsCloudContainer/Infrastructure/IImageGenerator.cs:
```diff
@@ -1,10 +1,12 @@
     internal interface IImageGenerator
     {
-        public Bitmap GenerateImage(Dictionary<string, Rectangle> cloud, Size imageSize);
+        public Bitmap GenerateImage(IEnumerable<Tag> cloud, ImageSettings settings);
```
Правда ли `settings` — ответственность метода `GenerateImage`?
Может лучше сложить это в саму зависимость `IImageGenerator`?

catsAndSuicide replied:
То есть в интерфейс добавить свойство, хранящее ImageSettings?

xackill replied:
Ммм... Нет, это будет суперстранно.
Тут речь шла про конструктор класса, который реализует `IImageGenerator`.

xackill commented on TagsCloudContainer/App/Program.cs:
```diff
@@ -1,9 +1,41 @@
+
+namespace TagsCloudContainer.App
 {
     internal class Program
```
Давай еще синхронизируем, что ты планируешь добавить в решение. Можно прямо по пунктам :)

catsAndSuicide replied:
1) Долго колебалась между консольным и графическим клиентом и решила все же графического реализовать.
2) Нужно фильтр междометий, предлогов и союзов добавить, но пока не разобралась, как это реализовать (определять, к какой части речи принадлежит слово, или можно заранее завести список слов, которые считаются скучными, и проверять слово на принадлежность к этому списку?)

xackill replied:
По поводу 2. в форммулировке задачи есть ссылка на [библиотеки](https://github.com/kontur-courses/di/blob/master/HomeExercise.md#%D0%B4%D0%BE%D0%BF%D0%BE%D0%BB%D0%BD%D0%B8%D1%82%D0%B5%D0%BB%D1%8C%D0%BD%D1%8B%D0%B5-%D1%81%D1%81%D1%8B%D0%BB%D0%BA%D0%B8), которые в этом помогут

catsAndSuicide committed:
split infrastructure into directories

catsAndSuicide committed:
fixed infrastructure (and relevant implementations):

- renamed:
FileReader -> DataReader, FontSizeGetter -> FontGetter
- moved dependencies from methods GenerateImage, GenerateCloud, ReadLines to relevant constructors

catsAndSuicide committed:
fixed Utils:

- fixed bug in RectangleExtensions
- removed GetAllDirections from DirectionsUtils

catsAndSuicide committed:
refactoring: using ValueTuple in GenerateCloud and File.ReadLines in TxtReader

catsAndSuicide committed:
changed infrastructure: in IWordFilter FilterOutBoringWords -> IsBoring, fixed SimpleWordFilter

added extensions for IEnumerable<string>, fixed FrequencyDictionaryGenerator

catsAndSuicide committed:
renamed DictionaryGenerator -> TextParserToFrequencyDictionary

catsAndSuicide committed:
1) added interfaces and classes in infrastructure for gui

2) added implementations: CircularCloudAction, PictureBoxImageHolder, MainForm

catsAndSuicide committed:
1) fixed infrastructure and appropriate implementations: CloudGenerator, CloudPainter (renamed ImageGenerator), CloudVisualizer

2) added enum CloudLayouterAlgorithm and class CloudLayouterFactory
3) fixed Program: now it use MainForm

catsAndSuicide committed:
added ImageSettingsAction (work only image size settings)

catsAndSuicide committed:
renamed ImageSettings to AppSettings, added classes for imageSizeSettings, fontSettings, palette

catsAndSuicide committed:
added paletteSettingsAction

catsAndSuicide committed:
added fontSettingsAction

catsAndSuicide committed:
added saveImageAction

catsAndSuicide committed:
added DataReaderFactory, openFileAction (for input file)

catsAndSuicide committed:
now imageHolder does not hold AppSettings, all dependencies of AppSettings moved to constructors of classes

catsAndSuicide committed:
deleted AppSettings, added ImageFormatSettings, ImageSizeSettings, InputSettings, LayouterAlgorithmSettings, OutputSettings (all are singletons)

catsAndSuicide committed:
added part-of-speech filter

catsAndSuicide committed:
now cloud rendering is in different thread

catsAndSuicide committed:
added LiteratureTextParser and ToInitialFormNormalizer

catsAndSuicide committed:
fixed MainForm: now ClientSize is ImageSize from settings

catsAndSuicide committed:
added docx reader

catsAndSuicide committed:
added jpeg and bmp formats to save image

catsAndSuicide committed:
program refactoring

catsAndSuicide committed:
renamed TextParserToFrequencyDictionary -> TextAnalyzer

catsAndSuicide committed:
internal, private -> public for tests

catsAndSuicide committed:
added tests

xackill commented:
**Общее замечания:**
- По коммитам стало значительно лучше в плане разбиения; есть еще некоторые шаги, которые можно было бы упразднить (DictionaryGenerator -> TextParserToFrequencyDictionary -> TextAnalyzer), но в целом уже гораздо проще ориентироваться в решении;
- А вот над чем еще стоит тут поработать &mdash; это над названиями коммитов: сейчас они описывают каждый шаг решения. Какие файлы изменились/переименовались я узнаю из diff-а, а вот общую логику вычленить из длинного сообщения сложно. Поэтому лучше всего подбирать емкие описание того, что произошло (и, возможно, комментарий **почему это произошло**, если нужен). я тут люблю такое правило (как и с названиями классов/переменных): если не получается назвать емко коммит, возможно, его нужно либо укрупнить с каким-то коммитом, либо наоборот разбить.

xackill commented on TagsCloudContainer/App/FrequencyDictionaryGenerator/FrequencyDictionaryGenerator.cs:
```diff
@@ -7,32 +7,27 @@ namespace TagsCloudContainer.App.FrequencyDictionaryGenerator
+            var notBoringWordsCount = notBoringWords.Count();
+            return notBoringWords
+                .GroupBy(word => word)
+                .ToDictionary(group => group.Key,
```
Да, стало читаться лучше, однако IEnumerable notBoringWords вычисляется дважды. Стоит сохранить результат перечисления в память (например, массивом): вычислять его дважды долго (а может и невозможно, если слова поступают из сетевого запроса, который отдает их ровно один раз).

xackill commented on TagsCloudContainerTests/DataReaderTests.cs:
```diff
@@ -0,0 +1,74 @@
+    internal class DataReaderTests
+    {
+        private readonly string txtFilePath = Path.Combine(Directory.GetCurrentDirectory(),
+            "..", "..", "..", "files", "input.txt");
```
Тест зависит от того, что находится ровно в определенной папке (у него сверху выкачен файл репозитория и он скомпилирован ровно туда, куда написано). Это приведет к нестабильному тесту (обычно их компилируют, архивируют папочку и уносят на сервер, где прогоняют тесты; на этом сервере может не быть репозитория).

Принято поступать так:
- У файла в репозитории (найди в контекстном меню что-то вроде File Properties) указать опцию "Copy to output directory: Copy always или Copy if newer". Тогда файл будет попадать в директорию выполнения теста (а к ней можно будет обратиться прямо в тестах)
- Это не замечание к исправлению, просто на будущее

xackill commented on TagsCloudContainer/App/Program.cs:
```diff
@@ -30,6 +32,8 @@ private static void Main(string[] args)
                 var imageHolder = new PictureBoxImageHolder(ImageSizeSettings.Instance, ImageFormatSettings.Instance);
                 var services = new ServiceCollection()
+                    .AddSingleton<Mysteam>(new Mysteam(Path.GetFullPath(Path.Combine(
+                        Directory.GetCurrentDirectory(), "..", "..", "..", "mystem.exe"))))
```
Но вот с mystem эту логику "где находится mystem.exe" лучше поправить: все-таки решение обычно загружается без репозитория сверху :)

xackill commented on TagsCloudContainer/App/TextParserToFrequencyDictionary/PartOfSpeechFilter.cs:
```diff
@@ -0,0 +1,23 @@
+
+        public bool IsBoring(string word)
+        {
+            return settings.BoringGramParts.Contains(mysteam.GetWords(word)[0].Lexems[0].GramPart);
```
Сейчас решение не обрабатывает ошибки (что если GetWords вернул пустой массив?) тут и в некоторых других местах. Конечно, предполагалась обработка `Exception-ов`, но в рамках заканчивающегося времени предлагаю оставить эту обработку для паттерна `Result<T>` в следующей задаче.

xackill commented on TagsCloudContainerTests/DataReaderTests.cs:
```diff
@@ -0,0 +1,74 @@
+
+namespace TagsCloudContainerTests
+{
+    internal class DataReaderTests
```
Лучше разделять тесты на конкретный класс в отдельный файл: т.е. отдельные файлы тестов для `DataReaderFactory`, `TxtReader` и т.д.: так будет проще понять, все ли покрыто тестами, где они есть, а где точно нужно что-то добавить.

xackill commented:
**Про добавить тесты:**
- Да, какие-то самые важные сценарии покрыты, но стоит их расширить. Сначала разбить 1 класс = 1 класс тестов, а потом пробежаться по сценариям уже существующих тестов, потом &mdash; тех, у которых еще нет.

xackill commented:
Оставленные комментарии стоит дофиксить уже в рамках TagCloud3.
Сейчас считаю, что с задачей DI ты справилась хорошо :)

Зачтено 👍 👍 👍
